[Setup] Make PostCode optional as few countries don't use them

vedanshujain opened the pull request:
### All Submissions:
* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

<!-- Mark completed items with an [x] -->

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

Some countries like Hong Kong, UAE etc don't use post codes. This commitswill make the Postcode / Zip field optional in setup wizard if this is the case.
If its not known whether a country uses postcode or not, default behavior would be to require the postcode.
Closes #23544 .

### How to test the changes in this Pull Request:

1. Go to WooCommerce Setup wizard: `wp-admin/index.php?page=wc-setup&step=store_setup`
2. Select a country with post code required like `India`. Leave the PostCode field empty and click on `Let's Go` button. An error message will be displayed informing that post code field is required.
3. Select a country with post code not required, like `United Arab Emirates`. Leave the `PostCode` field empty and click on `Let's Go` button. No error message will be shown and you will be able to proceed.

### Other information:
This increases the setup page size from 22.25KB to 22.91KB because of added locale data on frontend

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [x] Have you successfully run tests with your changes locally?

<!-- Mark completed items with an [x] -->

### Changelog entry

> Enter a summary of all changes on this Pull Request. This will appear in the changelog if accepted.

vedanshujain committed:
[Setup] Make PostCode optional as few countries don't use them

Some countries like Hong Kong, UAE etc don't use post codes, in which case this commit will make the field optional.
If its not known whether a country uses postcode or not, it will be required.

claudiosanches commented on includes/admin/class-wc-admin-setup-wizard.php:
```diff
@@ -199,6 +199,7 @@ public function enqueue_scripts() {
 			array(
 				'pending_jetpack_install' => $pending_jetpack ? 'yes' : 'no',
 				'states'                  => WC()->countries->get_states(),
+				'postcodes'               => wp_list_pluck( WC()->countries->get_country_locale(), 'postcode' ),
```
This throw several PHP notices like:

> PHP Notice:  Undefined index: postcode in wp-includes/class-wp-list-util.php on line 154

This happens because not all items inside the `WC()->countries->get_country_locale()` contains `postcode`.

vedanshujain replied:
Thanks for review, I missed this. Refactored to use a helper method instead.

claudiosanches reviewed (changes requested):
This need some refactor in order to work.

vedanshujain committed:
Use helper method instead of `wp_list_pluck` to avoid notices

In country locales, many times configuration for postcodes may not be defined. In these cases, using `wp_list_pluck` will throw notices.

We now use a small helper method to extract post code configuration from country locales.

claudiosanches reviewed (changes requested):
Still not working, I tested with a few countries that declares no support for postcode, but no luck.

vedanshujain commented:
That's weird, this seems to be working for me. Just to be clear, Postcode field will be required when either `required` is set to true, or its undefined. Postcode will not be required only when `required` is explicitly set to `false`.

So postcode will be required for countries like India (where `required` attribute in postcode object is not defined), and will be not be required for countries like Hong Kong where `required` is explicitly set to false.

@claudiosanches can you please tell me what countries are you testing with so that I can debug this better?

claudiosanches commented:
@vedanshujain sorry, another warning below, works just fine. Thanks.

claudiosanches reviewed (approved):
LGTM!
